go-controller/pkg/ovn/controller/egressfirewall/egressfirewall.go

```diff
@@ -385,32 +423,73 @@ func (oc *EFController) sync(key string) (updateErr error) {
 		shouldRemove = true
 	} else {
 		shouldAddUpdate = true
+		efExists = true
+		activeNetwork, err := oc.networkManager.GetActiveNetworkForNamespace(namespace)
+		if err != nil {
+			if util.IsInvalidPrimaryNetworkError(err) {
+				// Namespace requires P-UDN, but it does not exist. Remove EF config and surface error in status.
+				shouldRemove = true
+				shouldAddUpdate = false
+				updateErr = err
+			} else {
+				return fmt.Errorf("failed to get active network for egress firewall %s/%s namespace: %w",
+					namespace, efName, err)
+			}
+		} else {
+			ownerController := activeNetwork.GetNetworkName() + "-network-controller"
+			currentPgName = libovsdbutil.GetPortGroupName(getNamespacePortGroupDbIDs(namespace, ownerController))
+			currentSubnetsKey = subnetsKeyForNetInfo(activeNetwork)
+		}
 	}
 
 	oc.cache.LockKey(namespace)
 	defer oc.cache.UnlockKey(namespace)
 
-	if entry, ok := oc.cache.Load(namespace); ok {
-		if entry.stale {
-			shouldRemove = true
+	cachedPGName := ""
+	oldPGNameForCleanup := ""
+	cachedEntry, cachedEntryExists := oc.cache.Load(namespace)
+	if cachedEntryExists {
+		cachedPGName = cachedEntry.pgName
+		// If the port-group name changed (e.g., primary NAD switch), update EF first and then
+		// remove stale references from the old port group.
+		if currentPgName != "" && cachedEntry.pgName != "" && cachedEntry.pgName != currentPgName {
+			oldPGNameForCleanup = cachedEntry.pgName
 		}
 	}
 
+	// If nothing relevant changed since last apply, skip update. This is mainly for NAD-triggered
+	// reconciles (eg, netInfo changes that don't affect EF), but is guarded by EF resourceVersion
+	// to avoid skipping spec updates.
+	if efExists && shouldAddUpdate && !shouldRemove && cachedEntryExists &&
+		cachedEntry.efResourceVersion != "" &&
+		cachedEntry.efResourceVersion == ef.ResourceVersion &&
+		cachedEntry.pgName == currentPgName &&
+		cachedEntry.subnetsKey == currentSubnetsKey {
+		shouldAddUpdate = false
+	}
+
 	if shouldRemove {
 		// delete case
-		entry, ok := oc.cache.Load(namespace)
-		if !ok {
-			return nil
+		if cachedPGName == "" {
```

Comment:
reading the remove sequence again, don't we skip cleaning up old dnsNameResolver entries on EF update? we only do extra ACL cleanup I think